File: src/domain/entity/product.ts
import { ProductInfoDTO } from "../../application/dtos/product.dto";
export interface IAgeRange {
  minAge: number;
  maxAge: number;
}
export interface IProduct {
  _id: string;
  name: string;
  nameAr: string;
  description: string;
  descriptionAr: string;
  price: number;
  categoryId: string;
  createdBy: string;
  images: string[];
  sizes: string[];
  colors: string[];
  stock: number;
  brand?: string;
  ageRange?: IAgeRange;
  rating?: number;
  reviewsCount?: number;
  favoritesCount?: number;
  viewsCount?: number;
  isActive?: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export class ProductMapper {
  static toDTO(product: IProduct): ProductInfoDTO {
    return {
      _id: product._id,
      name: product.name,
      stock: product.stock,
      description: product.description,
      price: product.price,
      categoryId: product.categoryId,
      images: product.images,
      sizes: product.sizes,
      colors: product.colors,
      brand: product.brand,
      ageRange: product.ageRange,
      nameAr: product.nameAr,
      descriptionAr: product.descriptionAr,
      rating: product.rating,
      reviewsCount: product.reviewsCount,
      favoritesCount: product.favoritesCount,
      viewsCount: product.viewsCount,
      isActive: product.isActive,
    };
  }
}

File: Thawbuk-store/lib/domain/entities/product.dart
import 'package:equatable/equatable.dart';

class Product extends Equatable {
  final String id;
  final String name;
  final String description;
  final double price;
  final String categoryId;
  final String createdBy;
  final List<String> images;
  final List<String> sizes;
  final List<String> colors;
  final int stock;
  final String? brand;
  final AgeRange? ageRange;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Product({
    required this.id,
    required this.name,
    required this.description,
    required this.price,
    required this.categoryId,
    required this.createdBy,
    required this.images,
    required this.sizes,
    required this.colors,
    required this.stock,
    this.brand,
    this.ageRange,
    required this.createdAt,
    required this.updatedAt,
  });

  String get mainImage => images.isNotEmpty ? images.first : '';
  bool get isInStock => stock > 0;
  bool get hasMultipleSizes => sizes.length > 1;
  bool get hasMultipleColors => colors.length > 1;

  @override
  List<Object?> get props => [
        id,
        name,
        description,
        price,
        categoryId,
        createdBy,
        images,
        sizes,
        colors,
        stock,
        brand,
        ageRange,
        createdAt,
        updatedAt,
      ];
}

class AgeRange extends Equatable {
  final int minAge;
  final int maxAge;

  const AgeRange({
    required this.minAge,
    required this.maxAge,
  });

  bool isForAge(int age) => age >= minAge && age <= maxAge;

  String get displayText => '$minAge - $maxAge سنة';

  @override
  List<Object?> get props => [minAge, maxAge];
}